test/extended/openstack/cpms.go

```diff
@@ -0,0 +1,129 @@
+package openstack
+
+import (
+	"context"
+	"strconv"
+
+	g "github.com/onsi/ginkgo/v2"
+	o "github.com/onsi/gomega"
+	"github.com/stretchr/objx"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	e2e "k8s.io/kubernetes/test/e2e/framework"
+)
+
+var _ = g.Describe("[sig-installer][Suite:openshift/openstack] ControlPlane MachineSet", func() {
+
+	defer g.GinkgoRecover()
+
+	var ctx context.Context
+	var dc dynamic.Interface
+	var clientSet *kubernetes.Clientset
+	var controlPlaneMachineSets []objx.Map
+
+	g.BeforeEach(func() {
+		ctx = context.Background()
+
+		g.By("preparing openshift dynamic client")
+		cfg, err := e2e.LoadConfig()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		dc, err = dynamic.NewForConfig(cfg)
+		o.Expect(err).NotTo(o.HaveOccurred())
+		clientSet, err = e2e.LoadClientset()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		skipUnlessMachineAPIOperator(ctx, dc, clientSet.CoreV1().Namespaces())
+		controlPlaneMachineSets, err = getControlPlaneMachineSets(ctx, dc)
+		o.Expect(err).NotTo(o.HaveOccurred())
+	})
+
+	g.It("has role master", func() {
+		labels := controlPlaneMachineSets[0].Get("spec.template.machines_v1beta1_machine_openshift_io.metadata.labels").Data().(map[string]interface{})
+		o.Expect(labels[machineLabelRole]).To(o.Equal("master"), "unexpected or absent role label")
+	})
+
+	g.It("replica number corresponds to the number of Machines", func() {
+		replicaNumber, _ := strconv.Atoi(controlPlaneMachineSets[0].Get("spec.replicas").String())
+		o.Expect(getMasterMachines(ctx, dc)).
+			To(o.HaveLen(replicaNumber), "unexpected number of replicas for CPMS")
+	})
+
+	g.It("ProviderSpec template is correctly applied to Machines", func() {
+		masterMachines, err := getMasterMachines(ctx, dc)
+		o.Expect(err).NotTo(o.HaveOccurred(), "error fetching Masters machines")
+
+		cpmsFlavor := controlPlaneMachineSets[0].Get("spec.template.machines_v1beta1_machine_openshift_io.spec.providerSpec.value.flavor").String()
```

Comment:
just a suggestion: we could maybe use a `providerSpec` to keep things a bit more readable

```suggestion
		providerSpec := controlPlaneMachineSets[0].Get("spec.template.machines_v1beta1_machine_openshift_io.spec.providerSpec").ObjxMap()
		cpmsFlavor := providerSpec.Get("flavor").String()

```

Reply:
Done